cai_causal_graph/time_series_causal_graph.py

```diff
@@ -0,0 +1,673 @@
+        the variable name and the time lag.
+
+        :param identifier: The identifier of the node.
+        :param variable_type: The type of the variable.
+        :param meta: The metadata of the node.
+        :param node: The node to add.
+        :param kwargs: Additional keyword arguments.
+        :return: The added node.
+        """
+        node = super().add_node(identifier, variable_type=variable_type, meta=meta, node=node, **kwargs)
+        # populate the metadata for each node
+        vname, lag = get_variable_name_and_lag(node.identifier)
+        node.meta[self._meta_variable_name] = vname
+        node.meta[self._meta_time_name] = lag
+        return node
+
+    @reset_attributes
+    def replace_node(
+        self,
+        /,
+        node_id: NodeLike,
+        new_node_id: Optional[NodeLike] = None,
+        *,
+        variable_type: NodeVariableType = NodeVariableType.UNSPECIFIED,
+        meta: Optional[dict] = None,
+    ):
+        """Replace a node in the graph. See `cai_causal_graph.causal_graph.CausalGraph.replace_node` for more details."""
+        super().replace_node(node_id, new_node_id, variable_type=variable_type, meta=meta)
+
+    @reset_attributes
+    def delete_node(self, identifier: NodeLike):
+        """Delete a node from the graph. See `cai_causal_graph.causal_graph.CausalGraph.delete_node` for more details."""
+        super().delete_node(identifier)
+
+    @reset_attributes
+    def delete_edge(self, source: NodeLike, destination: NodeLike):
+        """Delete an edge from the graph. See `cai_causal_graph.causal_graph.CausalGraph.delete_edge` for more details."""
+        super().delete_edge(source, destination)
+
+    @reset_attributes
+    def add_edge(
+        self,
+        /,
+        source: Optional[NodeLike] = None,
+        destination: Optional[NodeLike] = None,
+        *,
+        edge_type: EDGE_T = EDGE_T.DIRECTED_EDGE,
+        meta: Optional[dict] = None,
+        edge: Optional[Edge] = None,
+        **kwargs,
+    ) -> Edge:
+        """
+        Add an edge to the graph. See `cai_causal_graph.causal_graph.CausalGraph.add_edge` for more details.
+
+        In addition to the `CausalGraph.add_edge` method, this method also populates the metadata of the nodes with
+        the variable name and the time lag.
+        """
+        edge = super().add_edge(source, destination, edge_type=edge_type, meta=meta, edge=edge, **kwargs)
+        # populate the metadata for each node
+        for node in self.get_nodes():
+            # extract variable name and lag from node name
+            vname, lag = get_variable_name_and_lag(node.identifier)
+            node.meta[self._meta_variable_name] = vname
+            node.meta[self._meta_time_name] = lag
+
+        return edge
+
+    @classmethod
+    def from_causal_graph(cls, causal_graph: CausalGraph, store_graphs: bool = False) -> TimeSeriesCausalGraph:
```

Comment:
why not just create an empty `TimeSeriesCausalGraph` and then `add_node` and `add_edge` for each edge/node in the provided causal graph? Seems like your solution is more complicated?

Reply:
I want to make sure to copy everything but if people say this is not necessary I can change it. @andrew-causalens any opinion?

Reply:
What would be missing from Ilya's suggestion? I guess `self._sepsets`? But the `self._skeleton`, `elf._nodes_by_identifier`, `self._edges_by_source`, and `self._edges_by_destination` would all be populated. Am I missing anything else?

Reply:
self._sepsets could be lost. But if you think all the machinery is not worth it I will remove it.

Reply:
If `self._sepsets` is the only thing that could be lost. Could do as Ilya said and then just copy that at the end. But @ilya-causalens do you have anything functionally wrong with current implementation or just think it is overly complex? Because if it works and ensures all protected members are preserved, I am inclined to say we leave it for now.

Reply:
If it works and there is a reason for it (like the `self._sepsets` not being copied) then let's keep as is of course! I just didn't know about it and it seemed pretty complex.